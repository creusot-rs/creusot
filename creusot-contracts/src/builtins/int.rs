use super::model::*;
use creusot_contracts_proc::*;
use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int;

impl PartialEq for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "eq_int"]
    fn eq(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ne_int"]
    fn ne(&self, _: &Int) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "le_int"]
    fn le(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ge_int"]
    fn ge(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "lt_int"]
    fn lt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "gt_int"]
    fn gt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    fn partial_cmp(&self, _: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }
}

impl From<u32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "u32_to_int"]
    fn from(_: u32) -> Self {
        panic!()
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "u32_model"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "i32_to_int"]
    fn from(_: i32) -> Self {
        panic!()
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "i32_model"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "usize_to_int"]
    fn from(_: usize) -> Self {
        panic!()
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "usize_model"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        panic!()
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "isize_model"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
