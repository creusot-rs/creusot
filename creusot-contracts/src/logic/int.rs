use super::model::*;
use crate::logic::*;

use creusot_contracts_proc::*;

use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int(*mut ());

impl WellFounded for Int {}

impl Int {
    #[trusted]
    #[logic]
    #[creusot::builtins = "int.Power.power"]
    pub fn pow(self, _: Int) -> Int {
        std::process::abort()
    }
}

impl From<u32> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "u32_to_int"]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn from(_: u32) -> Self {
        std::process::abort()
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "u32_model"]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "i32_to_int"]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn from(_: i32) -> Self {
        std::process::abort()
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "i32_model"]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "usize_to_int"]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: usize) -> Self {
        std::process::abort()
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "usize_model"]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        std::process::abort()
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[rustc_diagnostic_item = "isize_model"]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
