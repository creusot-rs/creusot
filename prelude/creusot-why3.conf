[prover_modifiers]
name = "Z3"
driver = "z3_nombqi_seq.drv"