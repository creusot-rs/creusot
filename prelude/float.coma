module Float32
  use int.Int
  use bv.Pow2int
  use real.Abs as Abs
  use real.FromInt as FromInt
  use real.Truncate as Truncate
  use real.RealInfix
  use export ieee_float.RoundingMode

  type t = < float 8 24 >

  (** {2 Part I - Public Interface}   *)

  constant eb : int = t'eb
  (** the number of bits in the exponent.  *)

  constant sb : int = t'sb
  (** the number of bits in the significand, including the hidden bit. *)

  axiom eb_gt_1: 1 < eb
  axiom sb_gt_1: 1 < sb

  (** {3 Constructors and Constants} *)

  constant zeroF     : t   (** +0.0 *)
  (* exp_bias = 2^(eb - 1) - 1 *)
  (* max_finite_exp = 2^sb - 2 - exp_bias = exp_bias *)
  (* max_significand = (2^eb + 2^eb - 1) * 2^(1-eb) *)
  (* max_value = (2^eb + 2^eb - 1) * 2^(1-eb) * 2 ^ max_finite_exp *)
  (* [m;x] = ( 1 + m * 2^(1-eb) ) * 2^( x - exp_bias ) *)
  (* max_value = [2^(eb-1); 2^sb - 2] *)

  (** {3 Operators} *)
  let addi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function addi mode t t : t
  let subi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function subi mode t t : t
  let muli (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function muli mode t t : t
  let divi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function divi mode t t : t
    (** The four basic operations, rounded in the given mode *)


  let abs (a : t) (ret (result : t)) = any
  function abs         t   : t   (** Absolute value *)

  let fma  (mode : mode) (a : t) (b : t) (c : t) (ret (result : t)) = any
  function fma  mode t t t : t   (** Fused multiply-add: x * y + z *)
  let sqrt (mode : mode) (a : t) (ret (result : t)) = any
  function sqrt mode   t   : t   (** Square root *)

  function neg (x : t) : t
  let neg (x : t) (ret (result : t)) = ret {neg x}

  function add (x : t) (y : t) : t = addi RNE x y
  let add (a : t) (b : t) (ret (result : t)) = ret {add a b}

  function sub (x : t) (y : t) : t = subi RNE x y
  let sub (a : t) (b : t) (ret (result : t)) = ret {sub a b}

  function mul (x : t) (y : t) : t = muli RNE x y
  let mul (a : t) (b : t) (ret (result : t)) = ret {mul a b}

  function div (x : t) (y : t) : t = divi RNE x y
  let div (a : t) (b : t) (ret (result : t)) = ret {div a b}

    (** Notations for operations in the default mode RNE *)

  function roundToIntegral mode t : t
  let roundToIntegral (mode : mode) (a : t) (ret (result : t) {result = roundToIntegral mode a}) = any
    (** Rounding to an integer *)

  function min       t t : t
  function max       t t : t

  predicate le t t
  predicate lt t t
  predicate eq t t
  predicate ge (x:t) (y:t) = le y x
  predicate gt (x:t) (y:t) = lt y x

  let le (a : t) (b : t) (ret (result : bool) { result = le a b }) = any
  let lt (a : t) (b : t) (ret (result : bool) { result = lt a b }) = any
  let eq (a : t) (b : t) (ret (result : bool) { result = eq a b }) = any
  let ge (a : t) (b : t) (ret (result : bool) { result = ge a b }) = any
  let gt (a : t) (b : t) (ret (result : bool) { result = gt a b }) = any
end
module Float64
  use int.Int
  use bv.Pow2int
  use real.Abs as Abs
  use real.FromInt as FromInt
  use real.Truncate as Truncate
  use real.RealInfix
  use export ieee_float.RoundingMode

  type t = < float 11 53 >

  (** {2 Part I - Public Interface}   *)

  constant eb : int = t'eb
  (** the number of bits in the exponent.  *)

  constant sb : int = t'sb
  (** the number of bits in the significand, including the hidden bit. *)

  axiom eb_gt_1: 1 < eb
  axiom sb_gt_1: 1 < sb

  (** {3 Constructors and Constants} *)

  constant zeroF     : t   (** +0.0 *)
  (* exp_bias = 2^(eb - 1) - 1 *)
  (* max_finite_exp = 2^sb - 2 - exp_bias = exp_bias *)
  (* max_significand = (2^eb + 2^eb - 1) * 2^(1-eb) *)
  (* max_value = (2^eb + 2^eb - 1) * 2^(1-eb) * 2 ^ max_finite_exp *)
  (* [m;x] = ( 1 + m * 2^(1-eb) ) * 2^( x - exp_bias ) *)
  (* max_value = [2^(eb-1); 2^sb - 2] *)

  (** {3 Operators} *)
  let addi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function addi mode t t : t
  let subi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function subi mode t t : t
  let muli (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function muli mode t t : t
  let divi (mode : mode) (a : t) (b : t) (ret (result : t)) = any
  function divi mode t t : t
    (** The four basic operations, rounded in the given mode *)


  let abs (a : t) (ret (result : t)) = any
  function abs         t   : t   (** Absolute value *)

  let fma  (mode : mode) (a : t) (b : t) (c : t) (ret (result : t)) = any
  function fma  mode t t t : t   (** Fused multiply-add: x * y + z *)
  let sqrt (mode : mode) (a : t) (ret (result : t)) = any
  function sqrt mode   t   : t   (** Square root *)

  function neg (x : t) : t
  let neg (x : t) (ret (result : t)) = ret {neg x}

  function add (x : t) (y : t) : t = addi RNE x y
  let add (a : t) (b : t) (ret (result : t)) = ret {add a b}

  function sub (x : t) (y : t) : t = subi RNE x y
  let sub (a : t) (b : t) (ret (result : t)) = ret {sub a b}

  function mul (x : t) (y : t) : t = muli RNE x y
  let mul (a : t) (b : t) (ret (result : t)) = ret {mul a b}

  function div (x : t) (y : t) : t = divi RNE x y
  let div (a : t) (b : t) (ret (result : t)) = ret {div a b}

    (** Notations for operations in the default mode RNE *)

  function roundToIntegral mode t : t
  let roundToIntegral (mode : mode) (a : t) (ret (result : t) {result = roundToIntegral mode a}) = any
    (** Rounding to an integer *)

  function min       t t : t
  function max       t t : t

  predicate le t t
  predicate lt t t
  predicate eq t t
  predicate ge (x:t) (y:t) = le y x
  predicate gt (x:t) (y:t) = lt y x

  let le (a : t) (b : t) (ret (result : bool) { result = le a b }) = any
  let lt (a : t) (b : t) (ret (result : bool) { result = lt a b }) = any
  let eq (a : t) (b : t) (ret (result : bool) { result = eq a b }) = any
  let ge (a : t) (b : t) (ret (result : bool) { result = ge a b }) = any
  let gt (a : t) (b : t) (ret (result : bool) { result = gt a b }) = any
end