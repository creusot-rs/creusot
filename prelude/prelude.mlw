module Prelude
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double

  type usize = uint64
  type isize = int64
  type opaque_ptr

  type borrowed 'a = { current : 'a ; final : 'a; }
  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final
  val borrow_mut (a : 'a) : borrowed 'a
  ensures { *result = a }
end
