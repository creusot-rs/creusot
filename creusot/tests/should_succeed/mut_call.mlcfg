
module CreusotContracts_Resolve_Impl1_Resolve_Stub
  type t
  use prelude.Borrow
  predicate resolve (self : borrowed t)
end
module CreusotContracts_Resolve_Impl1_Resolve_Interface
  type t
  use prelude.Borrow
  predicate resolve (self : borrowed t)
end
module CreusotContracts_Resolve_Impl1_Resolve
  type t
  use prelude.Borrow
  predicate resolve (self : borrowed t) =
     ^ self =  * self
  val resolve (self : borrowed t) : bool
    ensures { result = resolve self }
    
end
module MutCall_Kill_Interface
  use prelude.Borrow
  use mach.int.Int
  use mach.int.UInt32
  val kill [@cfg:stackify] [#"../mut_call.rs" 3 0 3 20] (_1' : borrowed uint32) : ()
end
module MutCall_Kill
  use prelude.Borrow
  use mach.int.Int
  use mach.int.UInt32
  clone CreusotContracts_Resolve_Impl1_Resolve as Resolve0 with
    type t = uint32
  let rec cfg kill [@cfg:stackify] [#"../mut_call.rs" 3 0 3 20] (_1' : borrowed uint32) : ()
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var _1 : borrowed uint32;
  {
    _1 <- _1';
    goto BB0
  }
  BB0 {
    _0 <- ();
    assume { Resolve0.resolve _1 };
    return _0
  }
  
end
module MutCall_Test_Interface
  val test [@cfg:stackify] [#"../mut_call.rs" 4 0 4 13] (_1' : ()) : ()
end
module MutCall_Test
  use mach.int.Int
  use mach.int.UInt32
  use prelude.Borrow
  clone CreusotContracts_Resolve_Impl1_Resolve as Resolve0 with
    type t = uint32
  clone MutCall_Kill_Interface as Kill0
  let rec cfg test [@cfg:stackify] [#"../mut_call.rs" 4 0 4 13] (_1' : ()) : () = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var a_1 : uint32;
  var _2 : ();
  var _3 : borrowed uint32;
  var _4 : borrowed uint32;
  {
    goto BB0
  }
  BB0 {
    a_1 <- ([#"../mut_call.rs" 5 16 5 18] (10 : uint32));
    _4 <- borrow_mut a_1;
    a_1 <-  ^ _4;
    _3 <- borrow_mut ( * _4);
    _4 <- { _4 with current = ( ^ _3) };
    _2 <- ([#"../mut_call.rs" 6 4 6 16] Kill0.kill _3);
    goto BB1
  }
  BB1 {
    assume { Resolve0.resolve _4 };
    _0 <- ();
    return _0
  }
  
end
