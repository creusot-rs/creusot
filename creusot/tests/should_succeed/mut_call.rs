fn kill(_: &mut u32) {}
fn test() {
    let mut a = 10;
    kill(&mut a);
}

fn main() {}
