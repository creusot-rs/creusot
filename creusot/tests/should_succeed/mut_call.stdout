module Type
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
end
module MutCall_Kill_Interface
  use prelude.Prelude
  use mach.int.Int
  use mach.int.UInt32
  val kill (_1 : borrowed uint32) : ()
end
module MutCall_Kill
  use prelude.Prelude
  use mach.int.Int
  use mach.int.UInt32
  let rec cfg kill (_1 : borrowed uint32) : () = 
  var _0 : ();
  var _1 : borrowed uint32;
  {
    _1 <- _1;
    goto BB0
  }
  BB0 {
    _0 <- ();
    assume { (fun x -> true) _1 };
    return _0
  }
  
end
module MutCall_Test_Interface
  val test () : ()
end
module MutCall_Test
  use mach.int.Int
  use mach.int.UInt32
  use prelude.Prelude
  clone MutCall_Kill_Interface as Kill0
  let rec cfg test () : () = 
  var _0 : ();
  var a_1 : uint32;
  var _2 : ();
  var _3 : borrowed uint32;
  var _4 : borrowed uint32;
  {
    goto BB0
  }
  BB0 {
    a_1 <- (10 : uint32);
    _4 <- borrow_mut a_1;
    a_1 <-  ^ _4;
    assume { (fun x -> true) a_1 };
    _3 <- borrow_mut ( * _4);
    _4 <- { _4 with current = ( ^ _3) };
    assume { (fun x -> true) _4 };
    _2 <- Kill0.kill _3;
    goto BB1
  }
  BB1 {
    assume { (fun x -> true) _3 };
    _0 <- ();
    return _0
  }
  
end
module MutCall_Main_Interface
  val main () : ()
end
module MutCall_Main
  let rec cfg main () : () = 
  var _0 : ();
  {
    goto BB0
  }
  BB0 {
    _0 <- ();
    return _0
  }
  
end
